infra/Pos/Slotting/Util.hs

```diff
@@ -137,40 +185,34 @@ onNewSlotImpl withLogging startImmediately action =
         -- REPORT:ERROR 'reportOrLogE' in 'onNewSlotImpl'
         reportOrLogE "Error occurred in 'onNewSlot' worker itself: " e
         delay =<< getNextEpochSlotDuration
-        onNewSlotImpl withLogging startImmediately action
+        onNewSlotImpl withLogging params action
 
 onNewSlotDo
-    :: OnNewSlot ctx m
-    => Bool -> Maybe SlotId -> Bool -> (SlotId -> m ()) -> m ()
-onNewSlotDo withLogging expectedSlotId startImmediately action = do
+    :: MonadOnNewSlot ctx m
+    => Bool -> Maybe SlotId -> OnNewSlotParams -> (SlotId -> m ()) -> m ()
+onNewSlotDo withLogging expectedSlotId onsp action = do
     curSlot <- waitUntilExpectedSlot
 
-    -- Note that the action can take more time than the slot
-    -- duration. In this case action for the next slot won't start
-    -- until the ongoing action finishes.
-    -- The caller should decide how to deal with long-running actions.
-    -- They can do whatever they want with 'action', the simplest thing is
-    -- 'void . fork' (which is likely not the best idea, for the reasons why
-    -- 'async' package should be used instead of 'fork'-ing threads manually.
-    --
-    -- There are few things to consider when one wants to run this action in a
-    -- separate thread every time:
-    -- 1. If more than one action is launched, they may use same resources
-    -- concurrently, so the code must account for it.
-    -- 2. If action hangs for some reason, there can be infinitely growing pool
-    -- of hanging actions with probably bad consequences.
-    --
-    -- See also: CSL-1606.
-    when startImmediately $ action curSlot
-
     let nextSlot = succ curSlot
     Timestamp curTime <- currentTimeSlotting
     Timestamp nextSlotStart <- getSlotStartEmpatically nextSlot
     let timeToWait = nextSlotStart - curTime
+
+    let applyTimeout a = case onspTerminationPolicy onsp of
+          NoTerminationPolicy -> a
+          NewSlotTerminationPolicy name ->
+              whenNothingM_ (timeout timeToWait a) $
+                  logWarning $ sformat
+                  ("Action "%stext%
+                   " hasn't finished before new slot started") name
```

Comment:
I'd write "was terminated because of timeout" to explicitly emphasize it was caller decision to kill it, not the action's one.